libvmi/os/linux/elfparse.c

```diff
@@ -0,0 +1,422 @@
+/* The LibVMI Library is an introspection library that simplifies access to
+ * memory in a target virtual machine or in a file containing a dump of
+ * a system's physical memory.  LibVMI is based on the XenAccess Library.
+ *
+ * Author: Thomas Dangl (thomas.dangl@posteo.de)
+ *
+ * This file is part of LibVMI.
+ *
+ * LibVMI is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or (at your
+ * option) any later version.
+ *
+ * LibVMI is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
+ * License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with LibVMI.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "private.h"
+#include "elfparse.h"
+
+status_t
+elfparse_validate_elf_image(
+    const uint8_t * const image,
+    size_t len)
+{
+    struct elf64_ehdr *elf = (struct elf64_ehdr *) image;
+
+    if (elf->e_ident[0] != 0x7F || elf->e_ident[1] != 'E'
+            || elf->e_ident[2] != 'L' || elf->e_ident[3] != 'F') {
+        dbprint(VMI_DEBUG_ELFPARSE, "--ELFPARSE: ELF header signature not found\n");
+        return VMI_FAILURE;
+    }
+
+    if (elf->e_phoff + sizeof(struct elf64_phdr) * elf->e_phnum > len) {
+        dbprint(VMI_DEBUG_ELFPARSE, "--ELFPARSE: program headers outside buffer\n");
+        return VMI_FAILURE;
+    }
+
+    if (elf->e_type != ET_EXEC && elf->e_type != ET_DYN) {
+        dbprint(VMI_DEBUG_ELFPARSE, "--ELFPARSE: unsupported file type\n");
+        return VMI_FAILURE;
+    }
+
+    if (elf->e_phentsize != sizeof(struct elf64_phdr)) {
+        dbprint(VMI_DEBUG_ELFPARSE, "--ELFPARSE: invalid program header size\n");
+        return VMI_FAILURE;
+    }
+
+    return VMI_SUCCESS;
+}
+
+status_t
+elfparse_get_image(
+    vmi_instance_t vmi,
+    const access_context_t *ctx,
+    size_t len,
+    const uint8_t * const image)
```

Comment:
I think it's strange that you declare this function to take `const` pointer for image just to immediately discard the `const` qualifier. Should adjust the function prototype to just take `uint8_t *image` as input.

Reply:
Okay, then we need to do the same for the PE format as well.